docs/writing-tests/making-a-testing-plan.md

```diff
@@ -0,0 +1,507 @@
+
+This algorithm exhibits different behavior depending on whether or not an item
+exists at the provided key. To test this thoroughly, we would write two tests:
+one test would verify that `null` is returned when there is no item at the
+provided key, and the other test would verify that an item we previously stored
+was correctly retrieved when we called the method with its name.
+
+### Sequence
+
+Even without branching, the interplay between sequential algorithm steps can
+suggest interesting test cases. If two steps have observable side-effects, then
+it can be useful to verify they happen in the correct order.
+
+Most of the time, step sequence is implicit in the nature of the
+algorithm--each step operates on the result of the step that precedes it, so
+verifying the end result implicitly verifies the sequence of the steps. But
+sometimes, the order of two steps isn't particularly relevant to the result of
+the overall algorithm. This makes it easier for implementations to diverge.
+
+There are many common patterns where step sequence is observable but not
+necessarily inherent to the correctness of the algorithm:
+
+- input validation (when an algorithm verifies that two or more input values
+  satisfy some criteria)
+- event dispatch (when an algorithm [fires two or more
+  events](https://dom.spec.whatwg.org/#concept-event-fire))
+- object property access (when an algorithm retrieves two or more property
+  values from an object provided as input)
+
+*Example:* The following text is an abbreviated excerpt of the algorithm that
+runs during drag operations (from [the HTML
+specification](https://html.spec.whatwg.org/multipage/dnd.html#dnd)):
+
+> [...]
+> 4. Otherwise, if the user ended the drag-and-drop operation (e.g. by
+>    releasing the mouse button in a mouse-driven drag-and-drop interface), or
+>    if the `drag` event was canceled, then this will be the last iteration.
+>    Run the following steps, then stop the drag-and-drop operation:
+>    1. If the [current drag
+>       operation](https://html.spec.whatwg.org/multipage/dnd.html#current-drag-operation)
+>       is "`none`" (no drag operation) [...] Otherwise, the drag operation
+>       might be a success; run these substeps:
+>       1. Let *dropped* be true.
+>       2. If the [current target
+>          element](https://html.spec.whatwg.org/multipage/dnd.html#current-target-element)
+>          is a DOM element, [fire a DND
+>          event](https://html.spec.whatwg.org/multipage/dnd.html#fire-a-dnd-event)
+>          named `drop` at it; otherwise, use platform-specific conventions for
+>          indicating a drop.
+>       3. [...]
+>    2. [Fire a DND
+>       event](https://html.spec.whatwg.org/multipage/dnd.html#fire-a-dnd-event)
+>       named `dragend` at the [source
+>       node](https://html.spec.whatwg.org/multipage/dnd.html#source-node).
+>    3. [...]
+
+A strong test suite will verify that the `drop` event is fired as specified,
```

Comment:
```suggestion
A good test suite will verify that the `drop` event is fired as specified,
```

(Strong feels too... strong.)

Reply:
Both "strong" and "good" are pretty generic. How does "thorough" sound to you?

Reply:
👍

Reply:
This is not addressed as far as I can tell.